frontend/src/components/explorer/gemBrowser/Reaction.vue

```diff
@@ -147,33 +113,8 @@ export default {
       meta: generateSocialMetaTags({ title, description }),
     };
   },
-  watch: {
-    '$route.params': 'setup',
-  },
-  async created() {
-    if (!this.model || this.model.short_name !== this.$route.params.model) {
-      const modelSelectionSuccessful = await this.$store.dispatch('models/selectModel', this.$route.params.model);
-      if (!modelSelectionSuccessful) {
-        this.modelNotFound = true;
-      }
-    }
-    this.setup();
-  },
   methods: {
-    async setup() {
-      this.showLoaderMessage = 'Loading reaction data';
-      this.rId = this.$route.params.id;
-      try {
-        const payload = { model: this.model, id: this.rId };
-        await this.$store.dispatch('reactions/getReactionData', payload);
-        this.componentNotFound = false;
-        this.showLoaderMessage = '';
-        await this.getRelatedReactions();
-      } catch {
-        this.componentNotFound = true;
-      }
-    },
-    async getRelatedReactions() {
+    async handleCallback() {
```

Comment:
Instead of triggering subsequent API calls via a callback, couldn't `query-component-action` be an array instead eg:
```
query-component-actions="['reactions/getReactionData','reactions/getRelatedReactionsForReaction']"
```

Reply:
Technically that would work I think but I'm not sure if it's an improvement because it doesn't really improve the intent. I might argue for renaming `handleCallback` to something like `handleComponentLoadedCallback`. What advantages do you see with having them in a list?

Reply:
I do see how the callback is needed for the metabolite page, but conceptually that is a different use case than loading data for the related reactions. As a developer I found this behaviour surprising, and less future proof - with a list it will be very easy to handle multi-request pages.

Reply:
I think currently it's clearer that there is one request needed for fetching a component and a subsequent function to handle stuff to do after the component has loaded. If it is a list then it is not clear the order of when things should be called. "Are the requests handled sequentially or concurrently?" Stylistically I think it would look awkward to have something like:

```javascript
await this.queryComponentActions[0];
if (this.queryComponentActions[1]) {
  await this.queryComponentActions[1];
}
```

And I think it would only be less and less intuitive how and when requests are being used if the list was to grow.

Reply:
The list would be handled concurrently, and I can't imagine it would be more than a handful of requests.

Reply:
I don't know if it makes sense for them to be handled concurrently, at least without rewriting a lot of the logic. For example if the URL contains an invalid metabolite id then it is unnecessary to send the `getRelatedMetabolites` request. Also since this is a refactoring issue I don't think we should go about changing the behavior of these components in this PR.

I was just saying that without looking into the `ComponentLayout` component, it wouldn't be straightforward for the developer to know how the requests are handled if they are provided in a list, whereas currently the intent of each binding is clearer.

In addition, I currently don't see any need to add a third function binding because all of the "post-component-loaded" logic can be wrapped into a single function, like we do in `Metabolite.vue`, where we make two separate requests. If we do ever want to change so we are sending all the requests concurrently, then all of them could be wrapped into the same function before passing into `ComponentLayout.vue` as well. In general I think it is difficult to try to be too future-proof because we don't know what changes we should accommodate for in the future.

Reply:
> I don't know if it makes sense for them to be handled concurrently, at least without rewriting a lot of the logic. For example if the URL contains an invalid metabolite id then it is unnecessary to send the `getRelatedMetabolites` request.

Users will rarely trigger such unnecessary requests, so it is acceptable to have to handle occasional bad requests.

> Also since this is a refactoring issue I don't think we should go about changing the behavior of these components in this PR.

I thought it was the right moment, because it is a refactoring issue, and specifically the refactoring that introduces the new behaviors of`query-component-action` and `@handleCallback`.

> I was just saying that without looking into the `ComponentLayout` component, it wouldn't be straightforward for the developer to know how the requests are handled if they are provided in a list, whereas currently the intent of each binding is clearer.

Right, now it's a first action triggering request that gets some data, and all other stuff in callback. While this is a separation in timings, I was arguing for separating the concerns: get all data requests (component-actions) together, and all non-data callbacks separately. If timing is an issue, one could always consider the first request in the array the "blocking" one, and all the rest concurrent. This would be controlled by an explicit behavior in the `setup` of `ComponentLayout` rather explicitly by using the `callback`.

> In addition, I currently don't see any need to add a third function binding because all of the "post-component-loaded" logic can be wrapped into a single function, like we do in `Metabolite.vue`, where we make two separate requests.

This wrapping together I find goes against the separation of concerns principle, and it is exactly what I think should be avoided. Perhaps it would be clearer if the props would be called `component-first-action` and `private-component-setup` or something like this, to communicate that the separation is about timing (but I still think the separation should be about concerns).

Anyway, it doesn't seem that I can change your mind about this, so I am resolving the conversation to signal that no changes are expected here.

Reply:
Although I was involved in some of the work here it's still @inghylt's PR and I don't want to assume authority over any of this. Was just curious about the list suggestion and provided my thoughts. So please unresolved this if you want to so others can get a chance to respond to your suggestion.

Reply:
Good suggestion. I do however feel like it will make things less clear and than they currently are, and I agree with the concerns @e0 raise. So I'm fine with keeping this conversation resolved